### PR TITLE
Refactor encoder benchmark

### PR DESCRIPTION
**Pull request by vkovacevicTT**
Created unified benchmark script for embedding models.

Added:
- BERT for sentence embedding ( Closes https://github.com/tenstorrent/tt-xla/issues/2182 Closes #638 )
- Qwen3-embedding-4B ( Closes https://github.com/tenstorrent/tt-xla/issues/2627)

All encoders are required to define and pass to benchmark 3 functions:
1. load_inputs
2. preprocess_input
3. process_output

This will be applied to other benchmarks as well #769 .

Moved bge_m3_encode to this benchmark, removed bge_m3_encode.py script and enabled device perf.

Moved warmup and fps measurement functions for vision models from utils.py to vision_benchmark.py.

Runs:
- [BERT](https://github.com/tenstorrent/tt-forge/actions/runs/20513190405/)
- [Qwen3-embedding-4B](https://github.com/tenstorrent/tt-forge/actions/runs/20513198551)
- [bge_m3_encode](https://github.com/tenstorrent/tt-forge/actions/runs/20513204146)

### CODE REVIEW

File: benchmark/tt-xla/encoder_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+
+    # Get tokenizer (required for encoder benchmarks)
+    if not hasattr(model_loader, "tokenizer") or model_loader.tokenizer is None:
+        raise ValueError("Model loader must provide a tokenizer for encoder benchmarks")
+    tokenizer = model_loader.tokenizer
+
+    if data_format == "bfloat16":
+        model = model.to(torch.bfloat16)
+    elif data_format == "float32":
+        model = model.to(torch.float32)
```
Comment (odjuricicTT) [inline benchmark/tt-xla/encoder_benchmark.py]: Is this needed? what is the default for these models? Can we check?

File: benchmark/tt-xla/encoder_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+        model_info: Model info string to determine sentence type
+    """
+    # Select sentences based on model type
+    model_info_lower = model_info.lower()
+    if "bert" in model_info_lower and "turkish" in model_info_lower:
+        base_sentences = BERT_SENTENCES
+    elif "qwen" in model_info_lower:
+        base_sentences = QWEN_SENTENCES
+    else:
+        base_sentences = MULTILINGUAL_SENTENCES
```
Comment (odjuricicTT) [inline benchmark/tt-xla/encoder_benchmark.py]: I don't like this, is there a reason that we cannot use the same inputs for all models?

If so, this should be extracted into the individual models tests, the same as with read_logtis_fn.

File: benchmark/tt-xla/encoder_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+    attention_mask = tokenized_inputs["attention_mask"].to(device)
+
+    # Get model outputs
+    outputs = model(input_ids=input_ids, attention_mask=attention_mask)
+
+    # Mean pooling: mask out padding tokens and compute mean
+    if hasattr(outputs, "last_hidden_state"):
+        token_embeddings = outputs.last_hidden_state
+    else:
+        token_embeddings = outputs[0]  # Last hidden state
```
Comment (odjuricicTT) [inline benchmark/tt-xla/encoder_benchmark.py]: This should be handles the same way as read_logits_fn.

The general idea is to keep the benchmarking code clean and non model specific. All things that are different for individual models should be specified in the model test and passed into the benchmarking functions.

File: benchmark/tt-xla/encoder_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+    framework_model, model_info, tokenizer = setup_model(model_loader, model_variant, data_format)
+
+    # TODO(vkovacevic): Determine encoding strategy based on model type
+    model_info_lower = model_info.lower()
+    if "qwen" in model_info_lower:
+        encode_fn = last_token_pool_encode
+        use_mean_pooling = False
+    else:
+        encode_fn = mean_pool_encode
+        use_mean_pooling = True
```
Comment (odjuricicTT) [inline benchmark/tt-xla/encoder_benchmark.py]: Same as the above comments about model specific IFs, this should be passed from model test.

File: benchmark/tt-xla/encoder_benchmark.py (hunk 3 above)
Comment (vkovacevicTT) [reply benchmark/tt-xla/encoder_benchmark.py]: Agreed, I will refactor this.

**Review by vvukomanTT — APPROVED** [review]

File: benchmark/tt-xla/bge_m3_encode.py [hunk 5]
```diff
@@ -28,7 +28,6 @@
 )
 
 os.environ["PJRT_DEVICE"] = "TT"
-os.environ["XLA_STABLEHLO_COMPILE"] = "1"
```
Comment (odjuricicTT) [inline benchmark/tt-xla/bge_m3_encode.py]: This file is no longer needed?

File: benchmark/tt-xla/encoder_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+            Signature: fn(outputs, model_inputs) -> embeddings.
+            This function should extract hidden states and apply the appropriate pooling.
+        required_pcc: Minimum PCC threshold for output validation
+        enable_weight_bfp8_conversion: Whether to enable bfp8 weight conversion
+        experimental_enable_permute_matmul_fusion: Whether to enable permute matmul fusion optimization
+
+    Returns:
+        Benchmark result containing performance metrics and model information
+    """
+    if training:
```
Comment (odjuricicTT) [inline benchmark/tt-xla/encoder_benchmark.py]: @claude Create an issue to fully remove this param from all benchmarks.

File: benchmark/tt-xla/encoder_benchmark.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+    print(f"PCC verification passed with PCC={pcc_value:.6f}")
+    evaluation_score = pcc_value
+
+    total_samples = batch_size * loop_count
+    samples_per_sec = total_samples / total_time
+
+    metadata = get_benchmark_metadata()
+
+    full_model_name = model_info_name
+    model_type = "Encoder, Text Embedding"
```
Comment (odjuricicTT) [inline benchmark/tt-xla/encoder_benchmark.py]: This will be changed https://github.com/tenstorrent/tt-forge/pull/768
So watch out when rebasing if the other PR gets merged first.

File: benchmark/tt-xla/utils.py [hunk 8]
```diff
@@ -308,115 +306,42 @@ def create_benchmark_result(
...
-    total_time_itterations = sum(itteration_times)
-    total_time = total_time_itterations + output_time
-
-    # Convert to seconds
-    total_time /= 1e9
-    return predictions, total_time
+    # Check if left padding was used (all sequences end with non-padding tokens)
+    left_padding = (attention_mask[:, -1].sum() == attention_mask.shape[0]).item()
+    if left_padding:
+        return hidden_states[:, -1]
```
Comment (odjuricicTT) [inline benchmark/tt-xla/utils.py]: Do we need this if? Do we execute both cases?

File: benchmark/tt-xla/encoders.py [hunk 9]
```diff
@@ -0,0 +1,444 @@
...
+            sentences,
+            padding="max_length",
+            truncation=True,
+            max_length=input_sequence_length,
+            return_tensors="pt",
+        ).items()
+    }
+
+    # Create output processing function
+    output_processor_fn = lambda out, inputs: apply_mean_pooling(out.last_hidden_state, inputs["attention_mask"])
```
[the remark this answers is not in the document]
Comment (vkovacevicTT) [reply benchmark/tt-xla/encoders.py]: I would leave it like this in case some models don't have `out.last_hidden_state`.

File: benchmark/tt-xla/encoder_benchmark.py (hunk 6 above)
Comment (vkovacevicTT) [reply benchmark/tt-xla/encoder_benchmark.py]: #769
